File: 317-shortest-distance-from-all-buildings/317-shortest-distance-from-all-buildings.go

const (
    emptyLand = iota
    house
    obstacle
)

// 317. Shortest Distance from All Buildings
// BFS from buildings to empty lands approach
// time: O(m^2 * n^2)
// space: O(m*n)

func shortestDistance(grid [][]int) int {
    numX, numY := len(grid), len(grid[0])
    // condition: 1 <= m, n <= 50
    // enqueue all the buildings
    var (
        houseCoord [2]int
        houseQueue [][2]int
    )
    for x := 0; x < numX; x++ {
        for y := 0; y < numY; y++ {
            if grid[x][y] == house {
                houseQueue = append(houseQueue, [2]int{x, y})
            }
        }
    }
    // constraint: There will be at least one building in the grid.
    totalHouses := len(houseQueue)
    
    distanceGrid := make([][]int, numX)
    numReachGrid := make([][]int, numX)
    for x := range distanceGrid {
        distanceGrid[x] = make([]int, numY)
        numReachGrid[x] = make([]int, numY)
    }
    numReachable := 1
    for len(houseQueue) != 0 {
        houseCoord, houseQueue = houseQueue[0], houseQueue[1:]
        bfs(grid, distanceGrid, numReachGrid, houseCoord, numReachable)
        numReachable++
    }
    const maxInt32 = 1 << 31 - 1
    minDist := maxInt32
    for x := 0; x < numX; x++ {
        for y := 0; y < numY; y++ {
            if numReachGrid[x][y] == totalHouses &&
            distanceGrid[x][y] < minDist {
                minDist = distanceGrid[x][y]
            }
        }
    }
    if minDist == maxInt32 {
        return -1
    }
    return minDist
}

func bfs(grid [][]int, distanceGrid [][]int,
         numReachGrid[][]int, houseCoord [2]int, numReachable int ) {
    
    numX, numY := len(grid), len(grid[0])
    dirs := [5]int{1, 0, -1, 0, 1}
    var (
        node [2]int
        queue [][2]int
    )
    dist := 0
    queue = append(queue, houseCoord)
    for len(queue) != 0 {
        for currLen := len(queue); currLen != 0; currLen-- {
            node, queue = queue[0], queue[1:]
            x, y := node[0], node[1]
            distanceGrid[x][y] += dist
            for i := 1; i < 5; i++ {
                newX := x + dirs[i-1]
                newY := y + dirs[i]
                if newX < 0 || newX >= numX || newY < 0 || newY >= numY {
                    continue
                }
                if grid[newX][newY] == house || grid[newX][newY] == obstacle {
                    continue
                }
                if numReachGrid[newX][newY] == numReachable - 1 {
                    queue = append(queue, [2]int{newX, newY})
                    numReachGrid[newX][newY] = numReachable
                }
            }
        }
        dist++
    }
}

File: 317-shortest-distance-from-all-buildings/317-shortest-distance-from-all-buildings.py

# 317. Shortest Distance from All Buildings
# time: O(m^2*n^2)
# space: O(m*n)

from collections import deque

class Solution:
    def shortestDistance(self, grid: List[List[int]]) -> int:
        def bfs(grid: List[List[int]], distance_grid: List[List[int]],
                reach_grid: List[List[int]], house: List[int], num_reach: int):

            num_x = len(grid)
            num_y = len(grid[0])
            dirs = [1, 0, -1, 0, 1]
            queue = deque([])
            queue.append(house)

            dist = 0
            while len(queue) != 0:
                curr_len = len(queue)
                for _ in range(curr_len):
                    node = queue.popleft()
                    x, y = node
                    distance_grid[x][y] += dist
                    for i in range(1, 5):
                        new_x = x + dirs[i-1]
                        new_y = y + dirs[i]
                        if new_x < 0 or new_x >= num_x or new_y < 0 or new_y >= num_y:
                            continue
                        if grid[new_x][new_y] == 1 or grid[new_x][new_y] == 2:
                            continue

                        if reach_grid[new_x][new_y] == num_reach - 1:
                            queue.append([new_x, new_y])
                            reach_grid[new_x][new_y] = num_reach
                dist += 1

        num_x, num_y = len(grid), len(grid[0])
        house_q = deque([])
        for x in range(num_x):
            for y in range(num_y):
                if grid[x][y] == 1:
                    house_q.append([x, y])
                    
        total = len(house_q) 
        distance_grid = [[0 for _ in range(num_y)] for _ in range(num_x)]
        reach_grid = [[0 for _ in range(num_y)] for _ in range(num_x)]
        
        min_dist  = float('inf')
        
        num_reach = 1
        while len(house_q) != 0:
            house = house_q.popleft()
            bfs(grid, distance_grid, reach_grid, house, num_reach)
            num_reach += 1
            
        print(distance_grid) 
        print(reach_grid)
        for x in range(num_x):
            for y in range(num_y):
                if reach_grid[x][y] == total and distance_grid[x][y] < min_dist:
                    min_dist = distance_grid[x][y]
        
        if min_dist == float('inf'):
            return -1
        
        return min_dist
        
            